app/controllers/cart_controller.rb

```diff
@@ -0,0 +1,100 @@
+class CartController < ApplicationController
+  before_action :ensure_user_cart_exists
+  before_action :set_cart
+
+  def show
+    @cart_items = @cart.cart_items.includes(:product)
+  end
+
+  def create
+    begin
+      product = Product.find(params[:product_id])
+      cart_item = @cart.cart_items.find_by(product: product)
+
+      if cart_item
+        cart_item.update(quantity: cart_item.quantity + 1)
+      else
+        @cart.cart_items.create(product: product, quantity: 1)
+      end
+
+      redirect_back(fallback_location: root_path, notice: "#{product.name} added to cart!")
+    rescue ActiveRecord::RecordNotFound
+      redirect_back(fallback_location: root_path, alert: "Product not found.")
+    end
+  end
+
+  def destroy
+    begin
+      cart_item = @cart.cart_items.find(params[:id])
+      product_name = cart_item.product.name
+      cart_item.destroy
+
+      redirect_to cart_path, notice: "#{product_name} removed from cart!"
+    rescue ActiveRecord::RecordNotFound
+      redirect_to cart_path, alert: "Cart item not found."
+    rescue => e
+      redirect_to cart_path, alert: "Failed to remove item from cart. Please try again."
+    end
+  end
+
+  def update
```

Comment:
Please move the update method before destroy to follow with rails default method's order
index show new edit create update destroy